modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/PlannerHelper.java

```diff
@@ -168,6 +191,84 @@ public static IgniteRel optimize(SqlNode sqlNode, IgnitePlanner planner) {
         }
     }
 
+    private static @Nullable IgniteRel tryOptimizeFast(SqlNode sqlNode, IgnitePlanner planner) {
+        if (!fastQueryOptimizationEnabled()) {
+            return null;
+        }
+
+        if (sqlNode instanceof SqlInsert) {
+            return tryOptimizeInsert((SqlInsert) sqlNode, planner);
+        }
+
+        return null;
+    }
+
+    private static @Nullable IgniteRel tryOptimizeInsert(SqlInsert insertNode, IgnitePlanner planner) {
+        SqlNode sourceNode = insertNode.getSource();
+
+        if (!(sourceNode instanceof SqlBasicCall && sourceNode.getKind() == SqlKind.VALUES)) {
+            // only simple `INSERT INTO ... VALUES (...)` statement is supported at the moment
+            return null;
+        }
+
+        List<SqlNode> rowConstructors = ((SqlBasicCall) sourceNode).getOperandList();
+
+        if (rowConstructors.size() != 1) {
+            // multirow insert currently are not supported by IgniteKeyValueModify 
+            return null;
+        }
+
+        IgniteSqlToRelConvertor converter = planner.sqlToRelConverter();
+        RelOptTable targetTable = converter.getTargetTable(insertNode);
+        IgniteTable igniteTable = targetTable.unwrap(IgniteTable.class);
+
+        assert igniteTable != null;
```

Comment:
at this point it must be `IgniteTable`. We don't have other tables but IgniteTable, and it's only relation that can be modified, thus after validation we don't have other options here. And to be clear: it's not a big difference between NPE and assertion error in production, both must be fixed asap. In this particular case we don't even loose stack trace, since NPE will be thrown on the very next line